### PR TITLE
Cave generator enhancements.

### PR DESCRIPTION
**Pull request by Zuhayr**
Essentially this changes the map generator to avoid using ChangeTurf and collapses the mining turfs into a single entity, using density as a toggle between wall/floor states. Testing on Europa indicated a speedup of almost 55 seconds - >60s to <5s.

### CODE REVIEW

File: code/modules/mining/mine_turfs.dm [hunk 1]
```diff
@@ -28,26 +30,108 @@
...
+	reconsider_lights()
+
+/turf/simulated/mineral/Entered(atom/movable/M as mob|obj)
+	. = ..()
+	if(istype(M,/mob/living/silicon/robot))
+		var/mob/living/silicon/robot/R = M
+		if(R.module)
+			if(istype(R.module_state_1,/obj/item/weapon/storage/bag/ore))
+				attackby(R.module_state_1,R)
+			else if(istype(R.module_state_2,/obj/item/weapon/storage/bag/ore))
```
Comment (GinjaNinja32) [inline code/modules/mining/mine_turfs.dm]:
```dm
for(var/obj/item/weapon/storage/bag/ore/O in list(R.module_state_1, R.module_state_2, R.module_state_3))
    attackby(O, R)
    return
```

File: code/modules/mining/mine_turfs.dm [hunk 2]
```diff
@@ -278,25 +415,17 @@
...
 	// Kill and update the space overlays around us.
+	var/list/step_overlays = list("n" = NORTH, "s" = SOUTH, "e" = EAST, "w" = WEST)
 	for(var/direction in step_overlays)
 		var/turf/space/T = get_step(src, step_overlays[direction])
 		if(istype(T))
 			T.overlays.Cut()
 			for(var/next_direction in step_overlays)
 				if(istype(get_step(T, step_overlays[next_direction]),/turf/simulated/mineral))
-					T.overlays += image('icons/turf/walls.dmi', "rock_side_[next_direction]")
+					T.overlays += image('icons/turf/walls.dmi', "rock_side", dir = step_overlays[next_direction])
```
Comment (GinjaNinja32) [inline code/modules/mining/mine_turfs.dm]: the non-numeric part of `step_overlays` is never used;
```dm
 for(var/direction in cardinal)
 	var/turf/space/T = get_step(src, direction)
 	if(istype(T))
 		T.overlays.Cut()
 		for(var/next_direction in cardinal)
 			if(istype(get_step(T, next_direction),/turf/simulated/mineral))
				T.overlays += image('icons/turf/walls.dmi', "rock_side", dir = next_direction)
```

File: code/modules/mining/mine_turfs.dm [hunk 3]
```diff
@@ -336,219 +464,33 @@
...
-				attackby(R.module_state_1,R)
-			else if(istype(R.module_state_2,/obj/item/weapon/storage/bag/ore))
-				attackby(R.module_state_2,R)
-			else if(istype(R.module_state_3,/obj/item/weapon/storage/bag/ore))
-				attackby(R.module_state_3,R)
-			else
-				return
\ No newline at end of file
+		mineral_name = pickweight(list("Uranium" = 5, "Platinum" = 5, "Iron" = 35, "Coal" = 35, "Diamond" = 1, "Gold" = 5, "Silver" = 5, "Phoron" = 10))
+	mineral_name = lowertext(mineral_name)
```
Comment (GinjaNinja32) [inline code/modules/mining/mine_turfs.dm]: uppercase names to pick from then `lowertext()`?

File: code/modules/mining/mine_turfs.dm (hunk 1 above)
Comment (Zuhayr) [inline code/modules/mining/mine_turfs.dm]: This is all existing code that has been collapsed, I was not intending to volunteer for a total refactor of this bollocks.

File: code/modules/mining/mine_turfs.dm (hunk 2 above)
Comment (Zuhayr) [inline code/modules/mining/mine_turfs.dm]: It was originally used as part of "rock_side", whoever fixed it to be directional didn't update the code thoroughly.

File: code/modules/mining/mine_turfs.dm (hunk 3 above)
Comment (Zuhayr) [inline code/modules/mining/mine_turfs.dm]: I have no idea.

File: code/game/objects/items/weapons/RCD.dm [hunk 4]
```diff
@@ -104,7 +104,7 @@
 		build_cost =  1
 		build_type =  "floor"
 		build_turf =  /turf/simulated/floor/airless
-	else if(!deconstruct && istype(T,/turf/simulated/floor/asteroid))
+	else if(!deconstruct && istype(T,/turf/simulated/mineral/floor))
```
Comment (Dennok) [inline code/game/objects/items/weapons/RCD.dm]: make RCD dont work.

Comment (Dennok) [inline code/game/objects/items/weapons/RCD.dm]: https://github.com/PolarisSS13/Polaris/pull/519/files#diff-dc31758bca74ab0a4951551cb88734d4R117
here need place 
else if(istype(T,/turf/simulated/floor) || istype(T,/turf/simulated/mineral/floor))

File: code/modules/mining/drilling/drill.dm [hunk 5]
```diff
@@ -74,11 +74,7 @@
 		return
 
 	//Drill through the flooring, if any.
-	if(istype(get_turf(src), /turf/simulated/floor/asteroid))
-		var/turf/simulated/floor/asteroid/T = get_turf(src)
-		if(!T.dug)
-			T.gets_dug()
-	else if(istype(get_turf(src), /turf/simulated/floor))
+	if(istype(get_turf(src), /turf/simulated/floor))
```
Comment (Dennok) [inline code/modules/mining/drilling/drill.dm]: why not place here this?
if(!istype(get_turf(src), /turf/simulated/mineral))

File: code/modules/mining/mine_items.dm [hunk 6]
```diff
@@ -235,7 +235,7 @@
 	var/obj/item/stack/flag/F = locate() in get_turf(src)
 
 	var/turf/T = get_turf(src)
-	if(!T || !istype(T,/turf/simulated/floor/asteroid))
+	if(!T || !istype(T,/turf/simulated/mineral))
```
Comment (Dennok) [inline code/modules/mining/mine_items.dm]: think better 
if(!T || !istype(T,/turf/simulated/mineral/floor))
or flag can be in wall?

File: code/modules/mining/mine_turfs.dm [hunk 7]
```diff
@@ -28,26 +32,113 @@
...
+		var/mob/living/silicon/robot/R = M
+		if(R.module)
+			for(var/obj/item/weapon/storage/bag/ore/O in list(R.module_state_1, R.module_state_2, R.module_state_3))
+				attackby(O, R)
+				return
+
+/turf/simulated/mineral/initialize()
+	if(prob(20))
+		overlay_detail = "asteroid[rand(0,9)]"
+	if(density)
```
Comment (Dennok) [inline code/modules/mining/mine_turfs.dm]: overlay_detail in walls?

File: code/modules/mining/mine_items.dm (hunk 6 above)
Comment (Zuhayr) [inline code/modules/mining/mine_items.dm]: Being in the wall requires admin jumping or a teleporter, it's an edge case that isn't worth checking for.

File: code/modules/mining/mine_turfs.dm (hunk 7 above)
Comment (Zuhayr) [inline code/modules/mining/mine_turfs.dm]: overlay_detail is set for all turfs and applied to floors in update_icon().

File: code/game/objects/items/weapons/RCD.dm (hunk 4 above)
Comment (Zuhayr) [inline code/game/objects/items/weapons/RCD.dm]: Noted, will check it out.

**Pull request merged**